cpu/esp32: add support for ESP32-S2

gschorcht opened the pull request:
### Contribution description

This PR is a split-off from PR #18235, which provides
- the support for ESP32-S2 SoC
- a common ESP32-S2 board definition
- the board definition for the ESP32S2-DevKit board.

This PR requires PRs #18502, #18503, #18504, #18505 and #18509 to be compilable in CI.

### Testing procedure

Beside of Green CI some basic tests should work (Of course, I have already tested all of them and many more):
- `tests/shell`
- `tests/periph_cpuid`
- `tests/periph_gpio`
- `tests/periph_hwrng`
- `tests/periph_i2c`
- `tests/periph_spi`
- `tests/periph_rtt`
- `tests/periph_timer`
- `examples/gnrc_networking` using `esp_wifi` as optional module.

### Issues/PRs references

Depends on #18502
Depends on #18503
Depends on #18504 
Depends on #18505 
Depends on #18509

benpicco commented on cpu/esp32/periph/spi.c:
```diff
@@ -312,6 +312,8 @@ void IRAM_ATTR spi_release(spi_t bus)
 #elif defined(CPU_FAM_ESP32C3)
 static const char* _spi_names[] = { "SPI", "FSPI"  };
+#elif defined(CPU_FAM_ESP32S2)
+static const char* _spi_names[] = { "SPI", "FSPI", "HSPI" };
```
Do the FSPI/HSPI names still make sense if they are only used for legacy reasons?

Doesn't have to be here, but would a follow up where we rename FSPI -> SPI2 and HSPI -> SPI2 make sense?

benpicco commented on cpu/esp32/include/periph_cpu_esp32s2.h:
```diff
@@ -0,0 +1,227 @@
+ * ESP32-S2 has two timer groups with two timers each, resulting in a total of
+ * four timers. Since one timer is used as system timer, up to three timers
+ * with one channel each can be used in RIOT as timer devices
+ * TIMER_DEV(0) ... TIMER_DEV(2).
```
Shouldn't `TIMER_NUMOF` be 3 then?

gschorcht replied:
If the hardware counter implementation for `periph/timer` is used (`MODULE_ESP_HW_COUNTER` is defined), there are three timer devices, but the third timer device uses a high-level interrupt that cannot be disabled. Therefore, the number of timer devices is limited to two, so only `TIMER_DEV(0)` and `TIMER_DEV(1)` can be used.

In case of the hardware timer implementation for `periph/timer` (`MODULE_ESP_HW_COUNTER` is not defined), the number of timers `TIMER_NUMOF` is derived from `SOC_TIMER_GROUP_TOTAL_TIMERS` here https://github.com/RIOT-OS/RIOT/blob/c908a60b7f03af55abc5e3337520fce633d318d6/cpu/esp32/include/periph_cpu.h#L659.

benpicco commented:
Awesome, works like a charm!
I had to patch two lines to get it to work though


Whitespace gets added to the variable 
```patch
--- a/cpu/esp32/bootloader/Makefile
+++ b/cpu/esp32/bootloader/Makefile
@@ -321,7 +321,8 @@ ESPTOOL ?= $(RIOTTOOLS)/esptools/esptool_v3.2.py
 # require to export these values.
 FLASH_MODE ?= dio   # ESP-IDF uses dio as default flash mode for the bootloader
 FLASH_FREQ ?= 40m   # lowest frequency all ESP32x SoC support
-FLASH_SIZE ?= 2     # smallesrt size all ESP32x SoC support
+# smallesrt size all ESP32x SoC support
+FLASH_SIZE ?= 2
```

otherwise I get

```
home/benpicco/dev/RIOT/dist/tools/esptools/esptool_v3.2.py --chip esp32s2 elf2image --flash_mode dio    \
  --flash_size 2     MB --flash_freq 40m    -o /home/benpicco/dev/RIOT/examples/gnrc_networking/bin/esp32s2-devkit/esp_bootloader/bootloader.bin /home/benpicco/dev/RIOT/examples/gnrc_networking/bin/esp32s2-devkit/esp_bootloader/bootloader.elf
usage: esptool elf2image [-h] [--output OUTPUT] [--version {1,2}] [--min-rev {0,1,2,3}] [--secure-pad]
                         [--secure-pad-v2] [--elf-sha256-offset ELF_SHA256_OFFSET] [--use_segments]
                         [--flash_freq {40m,26m,20m,80m}] [--flash_mode {qio,qout,dio,dout}]
                         [--flash_size FLASH_SIZE] [--spi-connection SPI_CONNECTION]
                         input
esptool elf2image: error: argument --flash_size/-fs: 2 is not a known flash size. Known sizes: 512KB, 256KB, 1MB, 2MB, 4MB, 2MB-c1, 4MB-c1, 8MB, 16MB
```

```patch
--- a/cpu/esp_common/esp-xtensa/thread_arch.c
+++ b/cpu/esp_common/esp-xtensa/thread_arch.c
@@ -286,7 +286,7 @@ void IRAM_ATTR thread_yield_higher(void)
         ets_soft_int_type = ETS_SOFT_INT_YIELD;
         WSR(BIT(ETS_SOFT_INUM), interrupt);
         critical_exit();
-#elif defined(CPU_FAM_ESP32)
+#elif defined(DPORT_CPU_INTR_FROM_CPU_0_REG)
         /* generate the software interrupt to switch the context */
         DPORT_WRITE_PERI_REG(DPORT_CPU_INTR_FROM_CPU_0_REG, DPORT_CPU_INTR_FROM_CPU_0);
 #else
```

benpicco commented:
Yea I don't know who thought it was a great idea that whitespace before a comment should be added to a variable, but we also had this problem with pkts recently: #18494

gschorcht commented:
> Yea I don't know who thought it was a great idea that whitespace before a comment should be added to a variable, but we also had this problem with pkts recently: #18494

Even if we can fix the problem for this variable in some way, the question remains why the default definition of `FLASH_SIZE` in `cpu/esp32/Makefile.include` is only not visible for the ESP32-S2 in `cpu/esp32/bootloader/Makefile` while it is for all other ESP32x SoCs.

benpicco commented:
But isn't the problem that only ESP32-S2 uses the default definition of `FLASH_SIZE`, so the others don't experience the problem with the trailing whitespace because they overwrite it?

gschorcht commented:
> But isn't the problem that only ESP32-S2 uses the default definition of `FLASH_SIZE`, so the others don't experience the problem with the trailing whitespace because they overwrite it?

I don't understand why the default definition in `cpu/esp32/bootloader/Makefile` is used for ESP32-S2 only. `FLASH_SIZE` is set for ESP32-S2 as for all other ESP32x SoCs here: https://github.com/RIOT-OS/RIOT/blob/5b6dff045a781c2fe4561dbce028199eb0d3b828/cpu/esp32/Makefile.include#L16-L19

benpicco commented:
Well what also works is 

```
--- a/cpu/esp32/bootloader/Makefile
+++ b/cpu/esp32/bootloader/Makefile
@@ -8,6 +8,7 @@ MODULE = esp_bootloader
 NO_AUTO_SRC = 1
 
 include $(RIOTBASE)/Makefile.base
+include $(RIOTBASE)/cpu/esp32/Makefile.include
 
 # Bootloader baudrate, set to the application defined one if any or the default
 # in serial.inc.mk
```

The file just does not get included by `cpu/esp32/bootloader/Makefile` otherwise

benpicco commented:
Ah I found the reason why

```patch
--- a/cpu/esp32/Makefile.include
+++ b/cpu/esp32/Makefile.include
@@ -14,9 +14,9 @@ else ifneq (,$(filter esp32c3 esp32s3,$(CPU_FAM)))
   export FLASH_SIZE ?= 2
   BOOTLOADER_POS = 0x0000
 else ifneq (,$(filter esp32s2,$(CPU_FAM)))
-  FLASH_MODE ?= qio
-  FLASH_FREQ ?= 80m
-  FLASH_SIZE ?= 4
+  export FLASH_MODE ?= qio
+  export FLASH_FREQ ?= 80m
+  export FLASH_SIZE ?= 4
   BOOTLOADER_POS = 0x1000
 else
   $(error Unkwnown ESP32x SoC variant (family))
```

gschorcht commented:
> Ah I found the reason why
> 
> ```diff
> --- a/cpu/esp32/Makefile.include
> +++ b/cpu/esp32/Makefile.include
> @@ -14,9 +14,9 @@ else ifneq (,$(filter esp32c3 esp32s3,$(CPU_FAM)))
>    export FLASH_SIZE ?= 2
>    BOOTLOADER_POS = 0x0000
>  else ifneq (,$(filter esp32s2,$(CPU_FAM)))
> -  FLASH_MODE ?= qio
> -  FLASH_FREQ ?= 80m
> -  FLASH_SIZE ?= 4
> +  export FLASH_MODE ?= qio
> +  export FLASH_FREQ ?= 80m
> +  export FLASH_SIZE ?= 4
>    BOOTLOADER_POS = 0x1000
>  else
>    $(error Unkwnown ESP32x SoC variant (family))
> ```

I looked at it 100 times and still missed this :worried:

gschorcht commented:
I had to rebase to fix it.

benpicco commented:
I still need this patch for it to compile

```patch
--- a/cpu/esp_common/esp-xtensa/thread_arch.c
+++ b/cpu/esp_common/esp-xtensa/thread_arch.c
@@ -286,7 +286,7 @@ void IRAM_ATTR thread_yield_higher(void)
         ets_soft_int_type = ETS_SOFT_INT_YIELD;
         WSR(BIT(ETS_SOFT_INUM), interrupt);
         critical_exit();
-#elif defined(CPU_FAM_ESP32)
+#elif defined(DPORT_CPU_INTR_FROM_CPU_0_REG)
         /* generate the software interrupt to switch the context */
         DPORT_WRITE_PERI_REG(DPORT_CPU_INTR_FROM_CPU_0_REG, DPORT_CPU_INTR_FROM_CPU_0);
 #else
```

benpicco reviewed (approved):
Please squash

gschorcht commented:
> Please squash

I'm a bit confused that I got hundreds of [makefile_broken](https://ci.riot-os.org/RIOT-OS/RIOT/18506/041c210a3edbe4f91b67990f275bdc48df06bd1a/output/makefile_broken.txt) errors when I tried to compile the PR in CI. I can't reproduce the error neither locally nor with `BUILD_IN_DOCKER=1` and `RIOT_CI_BUILD=1`.

benpicco commented on cpu/esp32/bootloader/sdkconfig_default_esp32s2.h:
```diff
@@ -0,0 +1,44 @@
+extern "C" {
+#endif
+
+#ifndef CONFIG_ESP32S3_DEFAULT_CPU_FREQ_MHZ
+#define CONFIG_ESP32S3_DEFAULT_CPU_FREQ_MHZ 160
+#endif
+
+#define CONFIG_ESP32S3_DEBUG_OCDAWARE 1
```
ESP32S2?

gschorcht replied:
> ESP32S2?

Good catch, a typical copy & paste error :worried: I fixed and squashed it directly.

benpicco commented:
Static tests are failing

     /data/riotbuild/boards/esp32s3-devkit/./Makefile.features:27: *** BOARD_VERSION is unknown.  Stop.

gschorcht commented:
> I still need this patch for it to compile
> 
> ```diff
> --- a/cpu/esp_common/esp-xtensa/thread_arch.c
> +++ b/cpu/esp_common/esp-xtensa/thread_arch.c
> @@ -286,7 +286,7 @@ void IRAM_ATTR thread_yield_higher(void)
>          ets_soft_int_type = ETS_SOFT_INT_YIELD;
>          WSR(BIT(ETS_SOFT_INUM), interrupt);
>          critical_exit();
> -#elif defined(CPU_FAM_ESP32)
> +#elif defined(DPORT_CPU_INTR_FROM_CPU_0_REG)
>          /* generate the software interrupt to switch the context */
>          DPORT_WRITE_PERI_REG(DPORT_CPU_INTR_FROM_CPU_0_REG, DPORT_CPU_INTR_FROM_CPU_0);
>  #else
> ```

Yeah, the change was part of of the initial PR #18235 https://github.com/RIOT-OS/RIOT/blob/7ebda1533df182b542e09c087044260373e5f26f/cpu/esp_common/thread_arch_xtensa.c#L291-L293 but got lost, when the files were restructered and the PR was rebased locally.

gschorcht commented:
> Static tests are failing
> 
> ```
>  /data/riotbuild/boards/esp32s3-devkit/./Makefile.features:27: *** BOARD_VERSION is unknown.  Stop.
> ```

Hm, why is it a problem for this PR and didn't pop up with PR #18421? Furthermore, I don't understand why this problem occurs, if `BOARD_VERSION` isn't defined, it is set at the beginning of `boards/esp32s3-devkit/Makefile.features` https://github.com/RIOT-OS/RIOT/blob/d118bb0f520379a9bdd7580fbeac67835b30146d/boards/esp32s3-devkit/Makefile.features#L1-L4

kaspar030 commented:
> Furthermore, I don't understand why this problem occurs, if `BOARD_VERSION` isn't defined, it is set at the beginning of `boards/esp32s3-devkit/Makefile.features`

I printed the value, when doing `make -Cexamples/hello-world info-boards-supported`, it is defined to `esp32s2-devkitm-1`.
That is in turn not handled in `boards/esp32s3-devkit/Makefile.features`, and it fails.

Maybe during dependency resolution, this needs to be cleared out? `make info-boards-supported` goes through the variables multiple times, and maybe this was set by a previous iteration (handling a previous board).

Somewhere here:

https://github.com/RIOT-OS/RIOT/blob/1583333b2d02b8d86d56548bf21b69d2685f89b5/makefiles/info-global.inc.mk#L41-L48

kaspar030 commented:
> Somewhere here:

(adding `undefine BOARD_VERSION` there fixes `make info-boards-supported`)

kaspar030 commented:
> Hm, why is it a problem for this PR and didn't pop up with PR #18421?

That didn't show up because there was only one user of BOARD_VERSION.

gschorcht commented:
> > Somewhere here:
> 
> (adding `undefine BOARD_VERSION` there fixes `make info-boards-supported`)

Hm, `BOARD_VERSION` isn't an official variable of RIOT's make system. I introduced it only for ESP32-S2-DevKit and ESP-S3-DevKit boards to avoid separate board definitions for dozens of variants of an identical board, differing only in the size of the integrated flash and the integrated SPI RAM. All these board variants are identical in their peripheral configuration so that separate board definitions would only blow up the number of boards in CI compilation a lot.

kaspar030 commented:
Sorry, this is one of the many quirks of the make based dependency resolution. It doesn't matter if the variable is "official", if one iteration sets a value that the next doesn't understand, the dep resolution fails.
I think it's a nice solution. But that variable needs to be cleared...

benpicco commented:
Now there is only a KConfig mismatch left for this to be merged :smiley:

gschorcht commented:
> Now there is only a [KConfig mismatch](https://ci.riot-os.org/RIOT-OS/RIOT/18506/34acec22a630a9849003421395dd166be546737c/output/compile/tests/kconfig_features/esp32s2-devkit:gnu.txt) left for this to be merged smiley

Hm, when I test it with
```
TEST_KCONFIG=1 BOARD=esp32s2-devkit make -j8 -C tests/shell info-features-provided
```
all provided features are the same as without `TEST_KCONFIG=1`.

benpicco commented on boards/esp32s2-devkit/Makefile.features:
```diff
@@ -0,0 +1,44 @@
+# default board version if not defined
+ifeq (,$(BOARD_VERSION))
+  BOARD_VERSION += esp32s2-devkitm-1
+endif
```
Why is this a list?
Shouldn't it be

```suggestion
BOARD_VERSION ?= esp32s2-devkitm-1
```

MrKevinWeiss commented on boards/esp32s2-devkit/Kconfig:
```diff
@@ -0,0 +1,65 @@
+
+    choice
+        bool "ESP32-S2-DevKit board version"
+        depends on TEST_KCONFIG
```
```suggestion

```

I think this is causing the error.  The board choice doesn't depend on the dependencies.

gschorcht replied:
Indeed, thanks.

gschorcht commented:
@benpicco I squashed the fix suggested by @MrKevinWeiss directly into commit 1862d4e6b80814d445c9c46d4c22dd48a1c4eacc. I added a new commit 18cf35b16f54655edafb6a3fb53e9fd448fdd644 to fix the default `BOARD_VERSION` setting.

gschorcht commented:
> @benpicco I squashed the fix suggested by @MrKevinWeiss directly into commit [1862d4e](https://github.com/RIOT-OS/RIOT/commit/1862d4e6b80814d445c9c46d4c22dd48a1c4eacc). I added a new commit [18cf35b](https://github.com/RIOT-OS/RIOT/commit/18cf35b16f54655edafb6a3fb53e9fd448fdd644) to fix the default `BOARD_VERSION` setting.

I fixed it for ESP32-S2 and ESP32-S3 instead of opening a separate PR for this very small change, even though this PR is not related to ESP32-S3.

gschorcht commented:
@benpicco Thanks for reviewing and merging.

gschorcht commented:
The next big challenge will be to upgrade ESP-IDF to version 5.0, which includes a lot of changes, also in the APIs. ESP-IDF 5.0 is required to support ESP32-C2 and ESP32-H2.

gschorcht commented:
@kaspar030 @MrKevinWeiss thanks for your ideas to solve the last remaining compile problems.
